# Project: Calculating lottery odds

In this project, we apply probability rules to calculate the chances of winning the lottery.

In [1]:
def factorial(n):
    output = 1
    for x in range (1,n+1):
        output *= x
    return output

def combinations(n,k):
    return factorial(n) / factorial(k) / factorial(n-k)

In [21]:
def one_ticket_probability(numbers):
    print('The probability of winning is {:.7f}%'.format(100/combinations(49,len(numbers))))
    print('There is a 1 in {:,} chances of winning'.format(int(combinations(49,len(numbers)))))

In [22]:
one_ticket_probability([1,2,3,4,5,6])

The probability of winning is 0.0000072%
There is a 1 in 13,983,816 chances of winning


The function that calculates the probability of one ticket takes in the chosen numbers as an input, but the chances are always the same regardless.

## Using historical data
We will also compare results with available data from drawings since 1982.

In [24]:
import pandas as pd
data = pd.read_csv('649.csv')
print(data.shape)
print(data.head(3))
print(data.tail(3))

(3665, 11)
   PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
0      649            1                0  6/12/1982               3   
1      649            2                0  6/19/1982               8   
2      649            3                0  6/26/1982               1   

   NUMBER DRAWN 2  NUMBER DRAWN 3  NUMBER DRAWN 4  NUMBER DRAWN 5  \
0              11              12              14              41   
1              33              36              37              39   
2               6              23              24              27   

   NUMBER DRAWN 6  BONUS NUMBER  
0              43            13  
1              41             9  
2              39            34  
      PRODUCT  DRAW NUMBER  SEQUENCE NUMBER  DRAW DATE  NUMBER DRAWN 1  \
3662      649         3589                0  6/13/2018               6   
3663      649         3590                0  6/16/2018               2   
3664      649         3591                0  6/20/2018              1

In [35]:
def extract_numbers(row):
    winners = set()
    for col in [4,5,6,7,8,9]:
        winners.add(row.iloc[col])
    return winners

data['winning_set'] = data.apply(extract_numbers, axis=1)

In [38]:
def check_historical_occurence(numbers, historical_data):
    numbers = set(numbers)
    print('These numbers appeared {} time(s) in the past'.format(sum(historical_data == numbers)))
    one_ticket_probability(list(numbers))

In [41]:
check_historical_occurence([3,41,11,12,43,14], data['winning_set'])

These numbers appeared 1 time(s) in the past
The probability of winning is 0.0000072%
There is a 1 in 13,983,816 chances of winning


In [40]:
check_historical_occurence([7,5,28,41,1,33], data['winning_set'])

These numbers appeared 0 time(s) in the past
The probability of winning is 0.0000072%
There is a 1 in 13,983,816 chances of winning


This new function checks the historical record for occurences of the numbers chosen. Since there are nearly 14 million combinations but only 3665 drawings since 1982, the odds of finding them are slim.

## Chances for multiple tickets
Let's now calculate the chances when playing multiple tickets... they shouldnt increase very much!

In [46]:
def multi_ticket_probability(nt):
    print('When playing {} tickets...'.format(nt))
    print('  The probability of winning is {:.7f}%'.format(nt*100/combinations(49,6)))
    print('  There is a 1 in {:,} chances of winning'.format(int(combinations(49,6))/nt))
    
for nt in [1, 10, 100, 10000, 1000000, 6991908, 13983816]:
    multi_ticket_probability(nt)

When playing 1 tickets...
  The probability of winning is 0.0000072%
  There is a 1 in 13,983,816.0 chances of winning
When playing 10 tickets...
  The probability of winning is 0.0000715%
  There is a 1 in 1,398,381.6 chances of winning
When playing 100 tickets...
  The probability of winning is 0.0007151%
  There is a 1 in 139,838.16 chances of winning
When playing 10000 tickets...
  The probability of winning is 0.0715112%
  There is a 1 in 1,398.3816 chances of winning
When playing 1000000 tickets...
  The probability of winning is 7.1511238%
  There is a 1 in 13.983816 chances of winning
When playing 6991908 tickets...
  The probability of winning is 50.0000000%
  There is a 1 in 2.0 chances of winning
When playing 13983816 tickets...
  The probability of winning is 100.0000000%
  There is a 1 in 1.0 chances of winning


The chance sof winning are multiplied by the number of tickets played.

## Winning with fewer numbers
Let's calculate the probability of getting smaller prizes.

In [49]:
def probability_less_6(w):
    combinations_possible = combinations(6,w)
    successful_outcomes = combinations_possible*combinations(49-w,6-w)
    print('The probability of winning is {:.7f}%'.format(successful_outcomes*100/combinations(49,6)))
    print('There is a 1 in {:,} chances of winning'.format(int(combinations(49,6)/successful_outcomes)))

In [50]:
probability_less_6(2)
probability_less_6(3)
probability_less_6(4)
probability_less_6(5)

The probability of winning is 19.1326531%
There is a 1 in 5 chances of winning
The probability of winning is 2.1710812%
There is a 1 in 46 chances of winning
The probability of winning is 0.1061942%
There is a 1 in 941 chances of winning
The probability of winning is 0.0018879%
There is a 1 in 52,969 chances of winning


In [ ]:
The chances of winning increase a lot wit